Retrying for actions + fixes

BartAdv opened the pull request:
- Uses same mechanics as retrying assertions
- Favours retrying whole action than just parts of it, as it may
  protects from StaleElementReference exceptions as well
- Fix for wait-and-click that waits for element to be enabled
- Unified retrying mechanics for smart-wait and other retries
- Fix assert-ui to use thread-local binding and timeout per assert-ui

BartAdv committed:
Retrying for actions + fixes

* Uses same mechanics as retrying assertions
* Favours retrying whole action than just parts of it, as it may
  protects from StaleElementReference exceptions as well
* Fix for wait-and-click that waits for element to be enabled
* Unified retrying mechanics for smart-wait and other retries
* Fix assert-ui to use thread-local binding and timeout per assert-ui

BartAdv commented on src/io/aviso/taxi_toolkit/utils.clj:
```diff
@@ -47,17 +48,22 @@
+          (throw ret)
+          (throw (TimeoutException.)))
+        :else (do
+                (Thread/sleep 17)
```
The sleep time is that small because if we've had an error here, it's quite possibly because an element had not yet been rendered/will be rerendered and the operation most likely will be successful fairly soon (approximating one frame). If the element is not yet there due to page being loaded, we're "sleeping" anyway because webdriver implicit wait so additional miliseconds won't give us anything.

clyfe replied:
weird sleep in retry, methinks

BartAdv replied:
You think it should be slower? Or just not be there at all?

clyfe replied:
Not there at all, failure should be overcome by a subsequent retry.

BartAdv replied:
It's probably unlikely that more than one such retry can occur in such short timespan, but I will leave the sleep here, as there's no point in being faster than browser refresh I think

clyfe replied:
Actually I think it's ok to wait a bit between failures, leave room for change.

BartAdv commented:
This changes a bit for fill-form that's been added recently, so I guess I'll summon @tlipski to review it as well.

tlipski commented:
No problem at all with the fill-form, the effect is quite similar. The `wait-until` call in line 110 of ui.clj might be perhaps dropped as well.

It might be good to also replace `retry` function as well (or just make it use the new function) if cleaning this up - right now there will be two timeout based functions and 0 retry-count based functions.

This change also makes me wonder if we shouldn't build regression tests for this library :)

BartAdv commented:
Well, that's why I pinged you. I think just running it against our test suites gives enough regression testing.

clyfe commented on src/io/aviso/taxi_toolkit/composite_assertions.clj:
```diff
@@ -24,25 +24,25 @@
+                     (if return
+                       (reset! result true)
+                       (throw (ex-info "forcing a retry" {}))))
+        return (binding [test/report spy-report]
```
This will play nice with test2junit?

BartAdv replied:
Yeah, test2junit itself uses binding. I now developed a habit of running `lein test2junit :only blahblah` locally as well.

tlipski commented:
Yes, that's probably sufficient, but on the other hand - this would happen after the coding and the feedback from particular test suites will not be available immediately. 
Just something to think about in the future.

clyfe commented on src/io/aviso/taxi_toolkit/ui.clj:
```diff
@@ -51,24 +55,30 @@
+  ;; TODO: this one is sometimes used as a checker in smart-wait/wait-until
+  ;; and sometimes as a standalone operation. The retrying here is not needed
+  ;; in case of smart-wait
+  (retrying
```
For the trick to work find+action need to be retried together.
Input is not an el-spec, it's an el afaict, won't work here. Maybe make the change deeper..

BartAdv replied:
Yeah, I think this particular helper kind of doesn't fit the API: we've got finders,waiters and actions that work on `el-spec` and assertions that work on `el`.

I think I've unnecesarily altered it, the `utils/retry` helper is still available, so I could just left it as it is, what do you think?

clyfe replied:
I'm thinking that whatever calls this should either use the spec, either retry both the find and this in _that_ place. Maybe there are few instances, in any case it's worth giving it a shot. O/w don't bother reverting.

BartAdv replied:
Yup, every call that uses it does something like `(tt/a-text (tt/$ ....))` so I guess we could change it for spec. It's just that would be breaking change:)

BartAdv replied:
And yeah, every other usage of it is something that should be fixed anyway.

BartAdv committed:
Use element specification for actions

This makes API consistent and more safe - using element references
without retrying may be unstable due to element becoming stale.

BartAdv commented:
OK, updated the API to avoid actions that take `element` as an argument (this should make it more consistent and safe). If you're OK with that change, let's merge it.

BartAdv committed:
Update changelog and readme

clyfe commented on src/io/aviso/taxi_toolkit/utils.clj:
```diff
@@ -47,17 +49,44 @@
+        (> (System/currentTimeMillis) (+ start timeout))
+        (if (instance? Throwable ret)
+          (throw ret)
+          (throw (TimeoutException.)))
```
Maybe this timeout ex should carry a message.

BartAdv replied:
Hm, yeah, I will check how it's then displayed in test summary, because when it's used from within `smart-wait` such info is added later.

BartAdv replied:
Checked, it's not needed:
- when called from `assert-ui` the action is retried once again after timeout to 'retrieve' original error
- when called from a waiter it's then wrapped by `smart-wait` and addtional info is added
